### PR TITLE
[13.0][MIG] link_tracker

### PR DESCRIPTION
**Pull request by Yasaie**
Migration of link_tracker
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/link_tracker/migrations/13.0.1.0/post-migration.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openupgradelib import openupgrade
+
+_unlink_by_xmlid = [
+    'link_tracker.view_link_tracker_form_stats',
```
Comment (MiquelRForgeFlow) [inline addons/link_tracker/migrations/13.0.1.0/post-migration.py]: why?

File: addons/link_tracker/migrations/13.0.1.0/post-migration.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+]
+
+
+def fill_link_tracker_code(env):
+    tracker_codes = env['link.tracker.code'].search([('code', '=', None)])
```
Comment (MiquelRForgeFlow) [inline addons/link_tracker/migrations/13.0.1.0/post-migration.py]:
```suggestion
    tracker_codes = env['link.tracker.code'].search([('code', '=', False)])
```

File: addons/link_tracker/migrations/13.0.1.0/pre-migration.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    ('link_tracker.view_link_tracker_filter', 'link_tracker.link_tracker_view_search'),
+    ('link_tracker.view_link_tracker_form', 'link_tracker.link_tracker_view_form'),
+    ('link_tracker.view_link_tracker_graph', 'link_tracker.link_tracker_view_graph'),
+    ('link_tracker.view_link_tracker_tree', 'link_tracker.link_tracker_view_tree'),
+]
+
+
+@openupgrade.migrate(use_env=True)
+def migrate(env, version):
+    openupgrade.rename_xmlids(env.cr, _xmlid_renames)
```
Comment (MiquelRForgeFlow) [inline addons/link_tracker/migrations/13.0.1.0/pre-migration.py]: no need: act_windows, menus, and views, when they are noupdate=0, then the deprecated ones will be deleted and the new ones will be created.

**Review by MiquelRForgeFlow — CHANGES_REQUESTED** [review]

File: addons/link_tracker/migrations/13.0.1.0/post-migration.py (hunk 2 above)
Comment (Yasaie) [reply addons/link_tracker/migrations/13.0.1.0/post-migration.py:10]: Isn't None equivalent to Null? why False?

File: addons/link_tracker/migrations/13.0.1.0/pre-migration.py (hunk 3 above)
Comment (Yasaie) [reply addons/link_tracker/migrations/13.0.1.0/pre-migration.py]: I thought it's better practice if I resolve all xmls deleted or renamed!
If not I'll skip noupdates of act_windows, menus and views

Comment (pedrobaeza) [reply addons/link_tracker/migrations/13.0.1.0/pre-migration.py]: Don't waste time, lines of code and so on for something that is done automatically. Here explicit is not better than implicit.

**Comment by Yasaie** [comment]
@MiquelRForgeFlow Thanks for reviews. changes made

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Hi, Payam, thanks for your contributions. OpenUpgrade migration scripts are the hardest thing to introduce in the OCA world, but it seems you are getting it.

I'm looking for the OCA CLA (https://odoo-community.org/page/cla) and I'm not able to find it. Did you send it?

**Comment by pedrobaeza** [comment]
Also, for being agile, would you allow me to push to your PR branches for adding/fixing things that I find in my review?

File: addons/link_tracker/migrations/13.0.1.0/openupgrade_analysis_work.txt [hunk 4]
```diff
@@ -0,0 +1,39 @@
+---Models in module 'link_tracker'---
+---Fields in module 'link_tracker'---
+link_tracker / link.tracker.click       / campaign_id (many2one)        : NEW relation: utm.campaign, isrelated: related, stored
+link_tracker / link.tracker.click       / click_date (date)             : DEL
+# NOTHING TO DO
```
Comment (pedrobaeza) [inline addons/link_tracker/migrations/13.0.1.0/openupgrade_analysis_work.txt]: It's good to always say why is `NOTHING TO DO` for future references or for the reviewers, and also leave an empty line for better readability when it's another "block", as we don't have syntax highlighting here in this file:

```suggestion
link_tracker / link.tracker.click       / campaign_id (many2one)        : NEW relation: utm.campaign, isrelated: related, stored
# NOTHING TO DO: Filled by standard update mechanism being a related. Possible improvement: being filled by SQL for performance.

link_tracker / link.tracker.click       / click_date (date)             : DEL
# NOTHING TO DO: This field was not used at all in previous version

```

File: addons/link_tracker/migrations/13.0.1.0/openupgrade_analysis_work.txt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+DEL ir.ui.view: link_tracker.view_link_tracker_click_tree
+DEL ir.ui.view: link_tracker.view_link_tracker_filter
+DEL ir.ui.view: link_tracker.view_link_tracker_form
+DEL ir.ui.view: link_tracker.view_link_tracker_graph
+DEL ir.ui.view: link_tracker.view_link_tracker_tree
+NEW ir.ui.view: link_tracker.link_tracker_click_view_search
+NEW ir.ui.view: link_tracker.utm_campaign_view_form
+NEW ir.ui.view: link_tracker.utm_campaign_view_kanban
+DEL ir.ui.view: link_tracker.view_link_tracker_form_stats
+# NOTHING TO DO
```
Comment (pedrobaeza) [inline addons/link_tracker/migrations/13.0.1.0/openupgrade_analysis_work.txt]:
```suggestion
# NOTHING TO DO: noupdate=0 records
```

File: addons/link_tracker/migrations/13.0.1.0/post-migration.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openupgradelib import openupgrade
+
+
+def fill_link_tracker_code(env):
```
Comment (pedrobaeza) [inline addons/link_tracker/migrations/13.0.1.0/post-migration.py:5]: Doing this in post-migration, will provoke a warning on update phase if there's any record without code. For amending it, you can:

- In pre, fill existing records without code with a fixed string like `openupgrade_temp_code`.
- In post, do this, but searching for `('code', '=', 'openupgrade_temp_code')`

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
I can do the changes if you allow me

**Comment by Yasaie** [comment]
Hi @pedrobaeza 

I have already sent the CLA request and I'm waiting to accept it soon.
Sure you can change as many as you can. But tell me what are they so I be more careful next time

Thank you

**Comment by Yasaie** [comment]
@pedrobaeza I have sent a PR [http://www.odoo.com/submit-pr](http://www.odoo.com/submit-pr) is this enough or I should send an email to cla AT odoo-community.org too?

**Comment by pedrobaeza** [comment]
That's Odoo core CLA, for contributing to Odoo core, not OCA CLA. Please see the link I sent you. Without it, it's not possible to merge this.

**Comment by Yasaie** [comment]
@pedrobaeza I sent Signed of OCA CLA. would you please confirm it
Thanks

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
CLA verified, thanks

**Pull request merged**